docs/sources/userguide/plugins.md

```diff
@@ -0,0 +1,159 @@
+Docker Plugins MVP
+==================
+
+Plugins provide a mechanism for hooking into the Docker engine to extend its behavior.
+
+HTTP over a local UNIX socket is used to communicate between Docker and its plugins.
+
+Plugins are distributed as containers.
+
+This is targeted at Docker 1.7.
+
+# Usage
+
+Plugins can be loaded using a special `docker plugin load` command, as follows:
+
+```
+$ docker plugin load -e $ARGS clusterhq/flocker-plugin $@
+[... fetching...]
+Plugin flocker:v0.1 loaded and registered with extension points:
+* volumes
+Plugin flocker activated.
+```
+
+This is really just syntactic sugar for the following:
+
+```
+$ docker run -d -e $ARGS \
+	-v /var/lib/docker/containers/<container_id>/plugin/:/var/run/docker/ \
+	-v /var/run/docker.sock:/var/run/docker.sock \
+	clusterhq/flocker-plugin $@
+```
+
+(The docker socket should only be mounted if the plugin is started with `--privileged`.)
+
+Loading a plugin forces it to always be loaded when Docker restarts (and Docker doesn't respond to API requests until it completes loading all its plugins).
+
+Docker then waits for the plugin to start listening on the socket (it polls the socket until it gets a successful response to an HTTP query to `/v1/handshake` on the socket, which returns with just a list of subsystems the plugin is interested in - response defined below).
+According to the type of plugin which is negotiated in the handshake, Docker registers the plugin to send it HTTP requests on certain events.
+
+Plugins should name themselves in the response, which should be in this format:
+
+```
+{
+ InterestedIn: ["volume"],
+ PluginName: "flocker"
+}
+```
+
+Every event should be sent to every plugin registered for that subsystem for now.
+Later the handshake can include more granular negotiation over which events it wants to receive.
+
+Other `plugin` subcommands can include `list`, `status`, `upgrade`, and `unload`.
+
+# Types of plugin
+
+Each plugin type defines a straightforward (MVP) protocol.
+
+## Volume
+
+We will start with this first because it is has the tiniest interface.
+
+The simplest of plugin types, `volume` provides a single request-response type on `POST /v1/volumes` (`POST ~= create`):
+
+**Request**
+
+```
+{HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
```

Comment:
We may want to consider responses that stop the current action.  For example, does the mount stop if the plug-in fails?  Do we allow the plug-in to stop the action at all?  What happens if it fails? Do we roll things back?
My initial guess is that a failed plug-in should stop the flow - otherwise users could get unexpected results - or worse, very bad results if one of the plug-ins is an auth-type check.  And therefore, if possible, I think things should rollback.